graph.hpp

```diff
@@ -0,0 +1,97 @@
+#ifndef GRAPH_H_INCLUDE
+#define GRAPH_H_INCLUDE
+#include<vector>
+#include<map>
+#include<set>
+#include<memory>
+
+typedef long long int64;
+
+class Graph {
+
+  // Nested class representing node in the graph to manage connections
+  class Node {
+    // Since this class is not used anywhere else, OK to make it a friend
+    friend class Graph;
+    // These represent incoming and outgoing edges.
+    // Reasons for having both sets:
+    // 1. When a node is deleted, this makes it finding out which edges to
+    //    delete much easier
+    // 2. This makes reverse a lot easier
+    std::shared_ptr<std::set<int64> > outgoing;
+    std::shared_ptr<std::set<int64> > incoming;
+    // A bunch of mutator methods to add/delete incoming/outgoing edges from
+    // node
+    void insertOutgoing(int64 to);
+    void insertIncoming(int64 from);
+    void eraseOutgoing(int64 to);
+    void eraseIncoming(int64 from);
+    // Two accessor methods. Return true if there exists an incoming/outgoing
+    // edge
+    bool containsEdgeFrom(int64 from);
+    bool containsEdgeTo(int64 to);
+    Node DeepCopy();
+    public:
+    Node() :
+      outgoing(new std::set<int64>()),
+      incoming(new std::set<int64>()) {}
+  };
+
+  // collection of nodes, mapping ids to nodes
+  std::map<int64, Node> nodemap;
+
+  // method to check if there is a certain node in the graph
+  inline bool Contains(int64 nodeID);
```

Comment:
oooooh i thought that that was ~the~ one thing where humans still had the stick but you're right. removed.